dbt/include/snowflake/macros/adapters.sql

```diff
@@ -150,21 +150,93 @@
 {% endmacro %}
 
 
-{% macro snowflake__list_relations_without_caching(schema_relation) %}
+{% macro snowflake__get_paginated_relations_array(max_iter, max_results_per_iter, max_total_results, schema_relation, total_count, watermark) %}
+
+  {% set paginated_relations = [] %}
+
+  {% for _ in range(0, max_iter) %}
+
+      {%- set paginated_sql -%}
+         show terse objects in {{ schema_relation }} limit {{ max_results_per_iter }} from '{{ watermark.table_name }}'
+      {%- endset -%}
+
+      {%- set paginated_result = run_query(paginated_sql) %}
+      {%- set paginated_n = (paginated_result | length) -%}
+
+      {#
+        terminating condition: if there are 0 records in the result we reached
+        the end exactly on the previous iteration
+      #}
+      {%- if paginated_n == 0 -%}
+        {%- break -%}
+      {%- endif -%}
+
+      {#
+        terminating condition: At some point the user needs to be reasonable with how
+        many objects are contained in their schemas. Since there was already
+        one iteration before attempting pagination, loop.index == max_iter means
+        the limit has been surpassed.
+      #}
+
+      {%- if loop.index == max_iter -%}
+        {%- set msg -%}
+           dbt will list a maximum of {{ max_total_results }} objects in schema {{ schema_relation }}.
+           Your schema exceeds this limit. Please contact support@getdbt.com for troubleshooting tips,
+           or review and reduce the number of objects contained.
+        {%- endset -%}
+
+        {% do exceptions.raise_compiler_error(msg) %}
+      {%- endif -%}
+
+      {%- do paginated_relations.append(paginated_result) -%}
+      {%- set total_count.sum = total_count.sum + paginated_n %}
+      {% set watermark.table_name = paginated_result.columns[1].values()[-1] %}
+
+      {#
+        terminating condition: paginated_n < max_results_per_iter means we reached the end
+      #}
+      {%- if paginated_n < max_results_per_iter -%}
+         {%- break -%}
+      {%- endif -%}
+    {%- endfor -%}
+
+  {{ return(paginated_relations) }}
+
+{% endmacro %}
+
+{% macro snowflake__list_relations_without_caching(schema_relation, max_iter=10, max_results_per_iter=10000) %}
+
+  {%- set max_total_results = max_results_per_iter * max_iter -%}
+
   {%- set sql -%}
-    show terse objects in {{ schema_relation }}
+    show terse objects in {{ schema_relation }} limit {{ max_results_per_iter }}
   {%- endset -%}
 
   {%- set result = run_query(sql) -%}
-  {% set maximum = 10000 %}
-  {% if (result | length) >= maximum %}
-    {% set msg %}
-      Too many objects in schema  {{ schema_relation }}! dbt can only get
-      information about schemas with fewer than {{ maximum }} objects.
-    {% endset %}
-    {% do exceptions.raise_compiler_error(msg) %}
+
+  {%- set n = (result | length) -%}
+  {%- set total_count = namespace(sum=n) -%}
```

Comment:
On the contrary, I think you did exactly the right thing with that naming convention.

I'm more confused about how you've used these names in the code. If I’m reading you right, line 222 should be `{% if total_count >= max_results_per_iter %}` (instead of `{% if n >= max_results_per_iter %}`) and all other references to total_count should be removed from `snowflake__get_paginated_relations_array` both its header (153), call (229), and implementation (192).